Add tool for exporting individual files to galaxy file source plugins.

jmchilton opened the pull request:
## What did you do? 
- Stole tool from EU for exporting individual files to file source plugins, working on making it a bit more general.


## Why did you make this change?
- It is a frequently requested change and it requires some support in Galaxy core for both testing and in the tool framework - so I don't think it belongs in the tool shed. This is a bit like the cloud send tool already so I think the precedence for adding data export tools that require Galaxy support.

## How to test the changes? 
- [X] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).

jmchilton committed:
Add tool for exporting individual files to Galaxy.

Stolen wholesale from EU. Trying to make it a bit more general though.

jmchilton commented on tools/data_export/export_remote.xml:
```diff
@@ -0,0 +1,60 @@
+                    #else:
+                        #set file_ext = $name.ext
+                    #end if
+                    '${name.element_identifier}.${file_ext}'
```
@jmchilton Add tool wrapper support for doing this natively. If the extension is already included don't duplicate it.

mvdbeek replied:
Something like https://github.com/galaxyproject/galaxy/pull/3220 ? I was going to pick this back up, but let me know if you have other plans.

jmchilton commented on tools/data_export/export_remote.xml:
```diff
@@ -0,0 +1,60 @@
+                    #if $name.ext == 'vcf_bgzip':
+                        #set file_ext = 'vcf.bz'
+                    #else:
+                        #set file_ext = $name.ext
```
@jmchilton Add datatype support this. Special case this in the datatype registry not in this tool.

mvdbeek committed:
Enable datatype aliases for stuff that has weird extensions in Galaxy

mvdbeek committed:
Add name_and_ext shortcut

mvdbeek commented on tools/data_export/export_remote.xml:
```diff
@@ -0,0 +1,55 @@
+                #end for
+                --names
+                #for $infile in $export_type.infiles:
+                    '$infile.name_and_ext'
```
- [x] We should also build a variant that is file system safe (but not so easy to find a good pattern, see https://github.com/galaxyproject/galaxy/pull/7937/files)
- [x] We should move this into a configfile to avoid shell limits
- [x] We should export MetadataFile objects as well (maybe configurable ?)

mvdbeek committed:
Provide files to export as json configfile

This avoids shell limits and quoting issues.

mvdbeek committed:
Sanitize exported filenames

Now can't include leading `.`, `/` and utf null, and is truncated at 255
characters. Duplicate filenames will raise an error.

mvdbeek committed:
Allow creating subdirectories

mvdbeek committed:
Allow creation of subdirectories on export

mvdbeek committed:
Mirror collection structure on exported filesystem

mvdbeek committed:
Simplify, fix char counting

mvdbeek committed:
Convert extensions

mvdbeek committed:
Implement export of FileParameter / MetadataFile files

mvdbeek committed:
Tweak test for clarity

mvdbeek committed:
Provide `$input.serialize()`, allow wrap_json to serialize collections

bgruening commented on lib/galaxy/tool_util/xsd/galaxy.xsd:
```diff
@@ -5307,7 +5307,9 @@ A contrived example of a tool that uses this is the test tool
-paths for data parameters set the ``data_style`` attribute to ``paths`` (see [inputs_as_json_with_paths.xml](https://github.com/galaxyproject/galaxy/blob/dev/test/functional/tools/inputs_as_json_with_paths.xml) for an example).
+paths for data or collection inputs set the ``data_style`` attribute to ``paths`` (see [inputs_as_json_with_paths.xml](https://github.com/galaxyproject/galaxy/blob/dev/test/functional/tools/inputs_as_json_with_paths.xml) for an example).
+To include a dictionary with staging paths, paths and metadata files set the ``data_style`` attribute to ``staging_path_and_source_path``.
+An example tool that uses ``staging_path_and_source_path`` is [inputs_as_json_with_staging_path_and_source_path.xml](https://github.com/galaxyproject/galaxy/blobl/dev/test/functional/tools/inputs_as_json_with_staging_path_and_source_path.xml)
```
`blobl` looks wrong.

mvdbeek replied:
```suggestion
An example tool that uses ``staging_path_and_source_path`` is [inputs_as_json_with_staging_path_and_source_path.xml](https://github.com/galaxyproject/galaxy/blob/dev/test/functional/tools/inputs_as_json_with_staging_path_and_source_path.xml)
```

bgruening commented on tools/data_export/export_remote.xml:
```diff
@@ -0,0 +1,67 @@
+<tool id="export_remote" name="Export datasets" version="0.1.0" profile="21.05">
+    <description>to remote files source</description>
+    <!-- TODO
```
here is still a TODO is that correct? Do we need to mark this tools as galaxy-env-tool?

mvdbeek replied:
There's no package yet, but that would allow removing the tool id from https://github.com/galaxyproject/galaxy/pull/11613/files#diff-17974c8048767aa536063088ba97dbc9c4475e7589002c3b2335f9e8490fc920R129

mvdbeek committed:
Fix typo in url
